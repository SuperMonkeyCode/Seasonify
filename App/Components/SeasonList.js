import React from 'react';
import { List, ListItem, Text, Grid, Col, Badge } from 'native-base';

import { Metrics, AtomicStyles } from '../Themes';

const SeasonList = ({ seasons }) => (
  <List>
    {seasons &&
      Object.keys(seasons).reverse().map(id => (
        <ListItem key={id}>
          <Grid>
            <Col size={Metrics.columnLeft} style={AtomicStyles.noFlex}>
              <Badge info>
                <Text>{`Season ${id}`}</Text>
              </Badge>
              <Text note>
                {seasons[id].episodes && `${seasons[id].episodes} ep. `}
                {seasons[id].year && `in ${seasons[id].year}`}
              </Text>
            </Col>
            <Col size={Metrics.columnRight}>
              {/* <Text>Renaud, Sylvain, Papa&Maman</Text> */}
            </Col>
          </Grid>
        </ListItem>
      ))}
  </List>
);

export default SeasonList;
