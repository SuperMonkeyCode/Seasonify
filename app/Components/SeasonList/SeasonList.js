/* @flow */
import { Button, Col, Grid, List, ListItem, Text } from 'native-base';
import { compose, pure, withHandlers } from 'recompose';
import React from 'react';

import { Metrics } from 'Themes';
import type { Seasons, ViewerInfo } from 'Types';
import { getContrastingTextColor } from 'Libs/Helpers';

import styles from './SeasonList.style';

type Props = {
  /* parent */
  seasons: Seasons,
  navigation: Object,
  /* connect */
  seasonViewings: { [season: string]: ViewerInfo[] },
  /* HOC */
  handleViewerPress: Function,
};

const enhance = compose(
  pure,
  withHandlers({
    handleViewerPress: ({ navigation }: Props) => (viewer: ViewerInfo) => {
      navigation.navigate('FriendDetailsPage', {
        friendId: viewer.friendId,
        friendName: viewer.name,
      });
    },
  }),
);

const getBadgeBgColor = (color: string) => ({ backgroundColor: color });

const getBadgeColor = (color: string) => ({ color: getContrastingTextColor(color) });

function SeasonList({ seasons, seasonViewings, handleViewerPress }: Props) {
  return (
    <List>
      {seasons &&
        objectValues(seasons).reverse().map(season =>
          (<ListItem key={season.id}>
            <Grid>
              <Col size={Metrics.columnLeft} style={styles.seasonHeader}>
                <Text style={styles.seasonTitle}>{`Season ${season.id}`}</Text>
                <Text note style={styles.seasonInfos}>
                  {season.episodes && `${season.episodes} ep.`}
                  {season.year && ` in ${season.year}`}
                </Text>
              </Col>
              <Col size={Metrics.columnRight} style={styles.viewerList}>
                {seasonViewings[season.id] &&
                  seasonViewings[season.id].map(viewer =>
                    (<Button
                      small
                      key={viewer.friendId}
                      style={{ ...styles.viewer, ...getBadgeBgColor(viewer.color) }}
                      onPress={() => handleViewerPress(viewer)}
                    >
                      <Text style={getBadgeColor(viewer.color)}>
                        {viewer.name}
                      </Text>
                    </Button>),
                  )}
              </Col>
            </Grid>
          </ListItem>),
        )}
    </List>
  );
}

export default enhance(SeasonList);
