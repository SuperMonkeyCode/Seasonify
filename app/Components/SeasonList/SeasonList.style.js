export default {
  seasonHeader: {
    flexDirection: 'column',
    justifyContent: 'flex-start',
    alignItems: 'flex-end', // not-working, so using alignSelf
  },
  seasonTitle: {
    alignSelf: 'flex-end',
  },
  seasonInfos: {
    alignSelf: 'flex-end',
  },
  viewerList: {
    flexDirection: 'row',
    flexWrap: 'wrap',
    justifyContent: 'flex-start',
    alignItems: 'flex-start', // not-working, so using alignSelf
    paddingHorizontal: 20,
  },
  viewer: {
    alignSelf: 'flex-start',
    marginHorizontal: 2,
    marginVertical: 2,
  },
};
