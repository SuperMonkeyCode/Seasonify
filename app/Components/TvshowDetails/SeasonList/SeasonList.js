/* @flow */
import { Button, Col, Grid, Icon, Left, List, ListItem, Text, View } from 'native-base';
import { TouchableOpacity } from 'react-native';
import { compose, pure, withHandlers } from 'recompose';
import React from 'react';

import { Metrics } from 'Themes';
import { getContrastingTextColor } from 'Libs/Helpers';
import Helptext from 'Components/Helptext';
import isEmpty from 'Libs/isEmpty';

import styles from './SeasonList.style';

type Props = {
  /* parent */
  tvshowId: string,
  seasons: Seasons,
  navigation: Object,
  /* connect */
  seasonViewings: { [season: string]: ViewerInfo[] },
  friends: Friends,
  viewingUpdate: Function,
  viewingUnview: Function,
  /* HOC */
  handleViewerPress: Function,
  handleViewerLongPress: Function,
  handleSeasonPress: Function,
};

const enhance = compose(
  withHandlers({
    handleViewerPress: ({ navigation }: Props) => (viewer: ViewerInfo) => {
      navigation.navigate('FriendDetailsPage', { friendId: viewer.friendId });
    },
    handleViewerLongPress: ({ tvshowId, viewingUnview }: Props) => (
      viewer: ViewerInfo,
    ) => {
      viewingUnview(tvshowId, viewer.friendId);
    },
    handleSeasonPress: ({ navigation, tvshowId, friends, viewingUpdate }: Props) => (
      seasonId: string,
    ) => {
      navigation.navigate('PickPage', {
        title: 'Who watched that season?',
        isMultiSelection: false,
        collection: friends,
        onSelect: friendId => {
          /* istanbul ignore next */
          viewingUpdate(tvshowId, friendId, seasonId);
        },
      });
    },
  }),
  pure,
);

const getBadgeBgColor = (color: string) => ({ backgroundColor: color });

const getBadgeColor = (color: string) => ({ color: getContrastingTextColor(color) });

function SeasonList({
  seasons,
  seasonViewings,
  handleViewerPress,
  handleViewerLongPress,
  handleSeasonPress,
}: Props) {
  let helpAlreadyDisplayed: boolean = false;
  const helpIfNeeded = () => {
    if (!helpAlreadyDisplayed && isEmpty(seasonViewings)) {
      helpAlreadyDisplayed = true;
      return <Helptext>Click on a season when someone viewed it</Helptext>;
    }
    return null;
  };

  return (
    <List>
      {seasons &&
        objectValues(seasons).reverse().map(season =>
          (<ListItem key={season.id} onPress={() => handleSeasonPress(season.id)}>
            <Grid>
              <Col size={Metrics.columnLeft}>
                <View style={styles.seasonHeader}>
                  {/* <Left style={styles.seasonButton}> */}
                  {/* <Icon name="add-circle" style={styles.icon} /> */}
                  {/* </Left> */}
                  <Text style={styles.seasonTitle}>{`Season ${season.id}`}</Text>
                  <Text note style={styles.seasonInfos}>
                    {season.episodes && `${season.episodes} ep.`}
                    {season.year && ` in ${season.year}`}
                  </Text>
                </View>
              </Col>
              <Col size={Metrics.columnRight} style={styles.viewerList}>
                {helpIfNeeded()}
                {seasonViewings[season.id] &&
                  seasonViewings[season.id].map(viewer =>
                    (<Button
                      small
                      key={viewer.friendId}
                      style={{ ...styles.viewer, ...getBadgeBgColor(viewer.color) }}
                      onPress={() => handleViewerPress(viewer)}
                      onLongPress={() => handleViewerLongPress(viewer)}
                    >
                      <Text style={getBadgeColor(viewer.color)}>
                        {viewer.name}
                      </Text>
                    </Button>),
                  )}
              </Col>
            </Grid>
          </ListItem>),
        )}
    </List>
  );
}

export default enhance(SeasonList);
