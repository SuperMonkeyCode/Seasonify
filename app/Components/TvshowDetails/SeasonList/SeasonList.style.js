export default {
  seasonHeader: {
    flexDirection: 'row',
    flexWrap: 'wrap',
    justifyContent: 'space-between',
    alignItems: 'flex-start',
  },
  seasonButton: {},
  seasonTitle: {
    flexGrow: 1,
    textAlign: 'right',
  },
  seasonInfos: {
    flexBasis: '100%',
    textAlign: 'right',
  },
  // icon: {
  //   fontSize: 20,
  // },
  viewerList: {
    flexDirection: 'row',
    flexWrap: 'wrap',
    justifyContent: 'flex-start',
    alignItems: 'flex-start',
    paddingHorizontal: 20,
  },
  viewer: {
    marginHorizontal: 2,
    marginVertical: 2,
  },
};
