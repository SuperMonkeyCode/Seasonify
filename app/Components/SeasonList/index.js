/* @flow */
import { connect } from 'react-redux';

import { friendSelectors } from 'State/friendState';
import { viewingActions, viewingSelectors } from 'State/viewingState';

import SeasonList from './SeasonList';

export default connect(
  (state, { tvshowId }) => ({
    seasonViewings: viewingSelectors.getSeasonsWithViewers(state, { tvshowId }),
    friends: friendSelectors.getFriends(state),
  }),
  {
    viewingUpdate: viewingActions.viewingUpdate,
  },
)(SeasonList);
