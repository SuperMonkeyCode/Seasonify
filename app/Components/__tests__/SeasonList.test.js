import React from 'react';

import SeasonList from '../SeasonList';

function setup(specificProps = {}) {
  const props = {
    seasons: {
      abc123: {
        id: 'abc123',
        episodes: 13,
        year: 2014,
      },
      xyz890: {
        id: 'xyz890',
        episodes: 13,
        year: 2015,
      },
    },
    ...specificProps,
  };
  const component = shallow(<SeasonList {...props} />);
  return {
    component,
    props,
  };
}

describe('Rendering', () => {
  it('should render', () => {
    const { component } = setup();
    expect(component).toMatchSnapshot();
  });
});
