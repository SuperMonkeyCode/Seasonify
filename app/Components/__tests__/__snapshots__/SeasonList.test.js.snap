// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rendering should render 1`] = `
<Styled(List)>
  <Styled(ListItem)>
    <GridNB>
      <ColumnNB
        size={38}
        style={
          Object {
            "flex": 0,
          }
        }
      >
        <Styled(Badge)
          info={true}
        >
          <Styled(Text)>
            Season xyz890
          </Styled(Text)>
        </Styled(Badge)>
        <Styled(Text)
          note={true}
        >
          13 ep. 
          in 2015
        </Styled(Text)>
      </ColumnNB>
      <ColumnNB
        size={62}
      />
    </GridNB>
  </Styled(ListItem)>
  <Styled(ListItem)>
    <GridNB>
      <ColumnNB
        size={38}
        style={
          Object {
            "flex": 0,
          }
        }
      >
        <Styled(Badge)
          info={true}
        >
          <Styled(Text)>
            Season abc123
          </Styled(Text)>
        </Styled(Badge)>
        <Styled(Text)
          note={true}
        >
          13 ep. 
          in 2014
        </Styled(Text)>
      </ColumnNB>
      <ColumnNB
        size={62}
      />
    </GridNB>
  </Styled(ListItem)>
</Styled(List)>
`;
